File: simulator/chemistry.py
"""Chemistry simulation - molecular assembly and reactions.

Models formation of molecules from atoms: water, amino acids,
nucleotides, and other biomolecules essential for life.
Chemical reactions are driven by energy, catalysis, and concentration.
"""
import math
import random
from dataclasses import dataclass, field
from typing import Optional

from simulator.constants import (
    K_B, BOND_ENERGY_COVALENT, BOND_ENERGY_HYDROGEN,
    BOND_ENERGY_VAN_DER_WAALS, AMINO_ACIDS,
)
from simulator.atomic import Atom, AtomicSystem


@dataclass
class Molecule:
    """A molecule: a collection of bonded atoms."""
    atoms: list = field(default_factory=list)
    name: str = ""
    formula: str = ""
    molecule_id: int = 0
    energy: float = 0.0
    position: list = field(default_factory=lambda: [0.0, 0.0, 0.0])
    is_organic: bool = False
    functional_groups: list = field(default_factory=list)

    _id_counter = 0

    def __post_init__(self):
        Molecule._id_counter += 1
        self.molecule_id = Molecule._id_counter
        if not self.formula and self.atoms:
            self._compute_formula()

    def _compute_formula(self):
        """Compute chemical formula from constituent atoms."""
        counts = {}
        for atom in self.atoms:
            sym = atom.symbol
            counts[sym] = counts.get(sym, 0) + 1

        # Standard chemistry ordering: C, H, then alphabetical
        parts = []
        for sym in ["C", "H"]:
            if sym in counts:
                n = counts.pop(sym)
                parts.append(f"{sym}{n}" if n > 1 else sym)
        for sym in sorted(counts):
            n = counts[sym]
            parts.append(f"{sym}{n}" if n > 1 else sym)
        self.formula = "".join(parts)

        # Check if organic
        has_c = any(a.atomic_number == 6 for a in self.atoms)
        has_h = any(a.atomic_number == 1 for a in self.atoms)
        self.is_organic = has_c and has_h

    @property
    def molecular_weight(self) -> float:
        return sum(a.mass_number for a in self.atoms)

    @property
    def atom_count(self) -> int:
        return len(self.atoms)

    def to_compact(self) -> str:
        return f"{self.name or self.formula}(mw={self.molecular_weight})"


class ChemicalReaction:
    """A chemical reaction with reactants, products, and energy."""

    def __init__(self, reactants: list, products: list,
                 activation_energy: float = 1.0,
                 delta_h: float = 0.0,
                 name: str = ""):
        self.reactants = reactants
        self.products = products
        self.activation_energy = activation_energy
        self.delta_h = delta_h  # Negative = exothermic
        self.name = name

    def can_proceed(self, temperature: float) -> bool:
        """Check if reaction can proceed at given temperature."""
        thermal_energy = K_B * temperature
        if thermal_energy <= 0:
            return False
        rate = math.exp(-self.activation_energy / thermal_energy)
        return random.random() < rate

    def to_compact(self) -> str:
        r = "+".join(self.reactants)
        p = "+".join(self.products)
        return f"{r}->{p}(Ea={self.activation_energy:.1f},dH={self.delta_h:.1f})"


class ChemicalSystem:
    """Manages molecular assembly and chemical reactions."""

    def __init__(self, atomic_system: AtomicSystem):
        self.atomic = atomic_system
        self.molecules: list[Molecule] = []
        self.reactions_occurred = 0
        self.water_count = 0
        self.amino_acid_count = 0
        self.nucleotide_count = 0

    def form_water(self) -> list[Molecule]:
        """Form water molecules: 2H + O -> H2O"""
        waters = []
        hydrogens = [a for a in self.atomic.atoms
                     if a.atomic_number == 1 and not a.bonds]
        oxygens = [a for a in self.atomic.atoms
                   if a.atomic_number == 8 and len(a.bonds) < 2]

        while len(hydrogens) >= 2 and oxygens:
            h1 = hydrogens.pop()
            h2 = hydrogens.pop()
            o = oxygens.pop()

            water = Molecule(
                atoms=[h1, h2, o],
                name="water",
                position=o.position[:],
            )
            waters.append(water)
            self.molecules.append(water)
            self.water_count += 1

            # Form bonds
            h1.bonds.append(o.atom_id)
            h2.bonds.append(o.atom_id)
            o.bonds.extend([h1.atom_id, h2.atom_id])

        return waters

    def form_methane(self) -> list[Molecule]:
        """Form methane: C + 4H -> CH4"""
        methanes = []
        carbons = [a for a in self.atomic.atoms
                   if a.atomic_number == 6 and not a.bonds]
        hydrogens = [a for a in self.atomic.atoms
                     if a.atomic_number == 1 and not a.bonds]

        while carbons and len(hydrogens) >= 4:
            c = carbons.pop()
            hs = [hydrogens.pop() for _ in range(4)]

            methane = Molecule(
                atoms=[c] + hs,
                name="methane",
                position=c.position[:],
            )
            methanes.append(methane)
            self.molecules.append(methane)

            for h in hs:
                c.bonds.append(h.atom_id)
                h.bonds.append(c.atom_id)

        return methanes

    def form_ammonia(self) -> list[Molecule]:
        """Form ammonia: N + 3H -> NH3"""
        ammonias = []
        nitrogens = [a for a in self.atomic.atoms
                     if a.atomic_number == 7 and not a.bonds]
        hydrogens = [a for a in self.atomic.atoms
                     if a.atomic_number == 1 and not a.bonds]

        while nitrogens and len(hydrogens) >= 3:
            n = nitrogens.pop()
            hs = [hydrogens.pop() for _ in range(3)]

            ammonia = Molecule(
                atoms=[n] + hs,
                name="ammonia",
                position=n.position[:],
            )
            ammonias.append(ammonia)
            self.molecules.append(ammonia)

            for h in hs:
                n.bonds.append(h.atom_id)
                h.bonds.append(n.atom_id)

        return ammonias

    def form_amino_acid(self, aa_type: str = "Gly") -> Optional[Molecule]:
        """Form an amino acid from available atoms.

        Amino acids need: C, H, O, N (and sometimes S).
        Simplified: requires 2C + 5H + 2O + 1N minimum (glycine).
        """
        carbons = [a for a in self.atomic.atoms
                   if a.atomic_number == 6 and not a.bonds]
        hydrogens = [a for a in self.atomic.atoms
                     if a.atomic_number == 1 and not a.bonds]
        oxygens = [a for a in self.atomic.atoms
                   if a.atomic_number == 8 and len(a.bonds) < 2]
        nitrogens = [a for a in self.atomic.atoms
                     if a.atomic_number == 7 and not a.bonds]

        if len(carbons) < 2 or len(hydrogens) < 5:
            return None
        if len(oxygens) < 2 or len(nitrogens) < 1:
            return None

        atoms = (
            [carbons.pop() for _ in range(2)]
            + [hydrogens.pop() for _ in range(5)]
            + [oxygens.pop() for _ in range(2)]
            + [nitrogens.pop()]
        )

        aa = Molecule(
            atoms=atoms,
            name=aa_type,
            position=atoms[0].position[:],
            is_organic=True,
            functional_groups=["amino", "carboxyl"],
        )
        self.molecules.append(aa)
        self.amino_acid_count += 1

        # Form internal bonds
        for i, atom in enumerate(atoms[1:], 1):
            atoms[0].bonds.append(atom.atom_id)
            atom.bonds.append(atoms[0].atom_id)

        return aa

    def form_nucleotide(self, base: str = "A") -> Optional[Molecule]:
        """Form a nucleotide (sugar + phosphate + base).

        Simplified: requires C5 + H10 + O7 + N2 + P minimum.
        We use available atoms to approximate.
        """
        carbons = [a for a in self.atomic.atoms
                   if a.atomic_number == 6 and not a.bonds]
        hydrogens = [a for a in self.atomic.atoms
                     if a.atomic_number == 1 and not a.bonds]
        oxygens = [a for a in self.atomic.atoms
                   if a.atomic_number == 8 and len(a.bonds) < 2]
        nitrogens = [a for a in self.atomic.atoms
                     if a.atomic_number == 7 and not a.bonds]

        # Simplified requirements
        if len(carbons) < 5 or len(hydrogens) < 8:
            return None
        if len(oxygens) < 4 or len(nitrogens) < 2:
            return None

        atoms = (
            [carbons.pop() for _ in range(5)]
            + [hydrogens.pop() for _ in range(8)]
            + [oxygens.pop() for _ in range(4)]
            + [nitrogens.pop() for _ in range(2)]
        )

        nuc = Molecule(
            atoms=atoms,
            name=f"nucleotide-{base}",
            position=atoms[0].position[:],
            is_organic=True,
            functional_groups=["sugar", "phosphate", "base"],
        )
        self.molecules.append(nuc)
        self.nucleotide_count += 1

        for i, atom in enumerate(atoms[1:], 1):
            atoms[0].bonds.append(atom.atom_id)
            atom.bonds.append(atoms[0].atom_id)

        return nuc

    def catalyzed_reaction(self, temperature: float,
                           catalyst_present: bool = False) -> int:
        """Run catalyzed reactions to form complex molecules."""
        formed = 0
        # Activation energy is lower with catalyst
        ea_factor = 0.3 if catalyst_present else 1.0

        thermal = K_B * temperature

        # Try to form amino acids
        if thermal > 0 and len(self.atomic.atoms) > 10:
            aa_prob = math.exp(-5.0 * ea_factor / (thermal + 1e-20))
            if random.random() < aa_prob:
                aa_type = random.choice(AMINO_ACIDS)
                if self.form_amino_acid(aa_type):
                    formed += 1
                    self.reactions_occurred += 1

        # Try to form nucleotides
        if thermal > 0 and len(self.atomic.atoms) > 19:
            nuc_prob = math.exp(-8.0 * ea_factor / (thermal + 1e-20))
            if random.random() < nuc_prob:
                base = random.choice(["A", "T", "G", "C"])
                if self.form_nucleotide(base):
                    formed += 1
                    self.reactions_occurred += 1

        return formed

    def molecule_census(self) -> dict:
        """Count molecules by type."""
        counts = {}
        for m in self.molecules:
            key = m.name or m.formula
            counts[key] = counts.get(key, 0) + 1
        return counts

    def to_compact(self) -> str:
        counts = self.molecule_census()
        count_str = ",".join(f"{k}:{v}" for k, v in sorted(counts.items()))
        return (f"CS[mol={len(self.molecules)} "
                f"H2O={self.water_count} "
                f"aa={self.amino_acid_count} "
                f"nuc={self.nucleotide_count} "
                f"rxn={self.reactions_occurred} {count_str}]")

File: apps/nodejs/chemistry.js
/**
 * Chemistry simulation - molecular assembly and reactions.
 *
 * Models formation of molecules from atoms: water, amino acids,
 * nucleotides, and other biomolecules essential for life.
 * Chemical reactions are driven by energy, catalysis, and concentration.
 */

import {
    K_B, BOND_ENERGY_COVALENT, BOND_ENERGY_HYDROGEN,
    BOND_ENERGY_VAN_DER_WAALS, AMINO_ACIDS,
} from './constants.js';

import { Atom, AtomicSystem } from './atomic.js';


// === Molecule ===

let moleculeIdCounter = 0;

export class Molecule {
    /** A molecule: a collection of bonded atoms. */
    constructor({
        atoms = [],
        name = "",
        formula = "",
        energy = 0.0,
        position = [0.0, 0.0, 0.0],
        isOrganic = false,
        functionalGroups = [],
    } = {}) {
        this.atoms = atoms;
        this.name = name;
        this.formula = formula;
        this.energy = energy;
        this.position = [...position];
        this.isOrganic = isOrganic;
        this.functionalGroups = [...functionalGroups];

        moleculeIdCounter++;
        this.moleculeId = moleculeIdCounter;

        if (!this.formula && this.atoms.length > 0) {
            this._computeFormula();
        }
    }

    _computeFormula() {
        const counts = {};
        for (const atom of this.atoms) {
            const sym = atom.symbol;
            counts[sym] = (counts[sym] || 0) + 1;
        }

        // Standard chemistry ordering: C, H, then alphabetical
        const parts = [];
        for (const sym of ["C", "H"]) {
            if (sym in counts) {
                const n = counts[sym];
                delete counts[sym];
                parts.push(n > 1 ? `${sym}${n}` : sym);
            }
        }
        for (const sym of Object.keys(counts).sort()) {
            const n = counts[sym];
            parts.push(n > 1 ? `${sym}${n}` : sym);
        }
        this.formula = parts.join("");

        // Check if organic
        const hasC = this.atoms.some(a => a.atomicNumber === 6);
        const hasH = this.atoms.some(a => a.atomicNumber === 1);
        this.isOrganic = hasC && hasH;
    }

    get molecularWeight() {
        return this.atoms.reduce((sum, a) => sum + a.massNumber, 0);
    }

    get atomCount() {
        return this.atoms.length;
    }

    toCompact() {
        return `${this.name || this.formula}(mw=${this.molecularWeight})`;
    }
}

/** Reset molecule ID counter (useful for tests). */
export function resetMoleculeIdCounter() {
    moleculeIdCounter = 0;
}


// === ChemicalReaction ===

export class ChemicalReaction {
    /** A chemical reaction with reactants, products, and energy. */
    constructor(reactants, products, activationEnergy = 1.0, deltaH = 0.0, name = "") {
        this.reactants = reactants;
        this.products = products;
        this.activationEnergy = activationEnergy;
        this.deltaH = deltaH;  // Negative = exothermic
        this.name = name;
    }

    /** Check if reaction can proceed at given temperature. */
    canProceed(temperature) {
        const thermalEnergy = K_B * temperature;
        if (thermalEnergy <= 0) return false;
        const rate = Math.exp(-this.activationEnergy / thermalEnergy);
        return Math.random() < rate;
    }

    toCompact() {
        const r = this.reactants.join("+");
        const p = this.products.join("+");
        return `${r}->${p}(Ea=${this.activationEnergy.toFixed(1)},dH=${this.deltaH.toFixed(1)})`;
    }
}


// === ChemicalSystem ===

export class ChemicalSystem {
    /** Manages molecular assembly and chemical reactions. */
    constructor(atomicSystem) {
        this.atomic = atomicSystem;
        this.molecules = [];
        this.reactionsOccurred = 0;
        this.waterCount = 0;
        this.aminoAcidCount = 0;
        this.nucleotideCount = 0;
    }

    /** Form water molecules: 2H + O -> H2O */
    formWater() {
        const waters = [];
        const hydrogens = this.atomic.atoms.filter(
            a => a.atomicNumber === 1 && a.bonds.length === 0
        );
        const oxygens = this.atomic.atoms.filter(
            a => a.atomicNumber === 8 && a.bonds.length < 2
        );

        while (hydrogens.length >= 2 && oxygens.length > 0) {
            const h1 = hydrogens.pop();
            const h2 = hydrogens.pop();
            const o = oxygens.pop();

            const water = new Molecule({
                atoms: [h1, h2, o],
                name: "water",
                position: [...o.position],
            });
            waters.push(water);
            this.molecules.push(water);
            this.waterCount++;

            // Form bonds
            h1.bonds.push(o.atomId);
            h2.bonds.push(o.atomId);
            o.bonds.push(h1.atomId, h2.atomId);
        }

        return waters;
    }

    /** Form methane: C + 4H -> CH4 */
    formMethane() {
        const methanes = [];
        const carbons = this.atomic.atoms.filter(
            a => a.atomicNumber === 6 && a.bonds.length === 0
        );
        const hydrogens = this.atomic.atoms.filter(
            a => a.atomicNumber === 1 && a.bonds.length === 0
        );

        while (carbons.length > 0 && hydrogens.length >= 4) {
            const c = carbons.pop();
            const hs = [];
            for (let i = 0; i < 4; i++) hs.push(hydrogens.pop());

            const methane = new Molecule({
                atoms: [c, ...hs],
                name: "methane",
                position: [...c.position],
            });
            methanes.push(methane);
            this.molecules.push(methane);

            for (const h of hs) {
                c.bonds.push(h.atomId);
                h.bonds.push(c.atomId);
            }
        }

        return methanes;
    }

    /** Form ammonia: N + 3H -> NH3 */
    formAmmonia() {
        const ammonias = [];
        const nitrogens = this.atomic.atoms.filter(
            a => a.atomicNumber === 7 && a.bonds.length === 0
        );
        const hydrogens = this.atomic.atoms.filter(
            a => a.atomicNumber === 1 && a.bonds.length === 0
        );

        while (nitrogens.length > 0 && hydrogens.length >= 3) {
            const n = nitrogens.pop();
            const hs = [];
            for (let i = 0; i < 3; i++) hs.push(hydrogens.pop());

            const ammonia = new Molecule({
                atoms: [n, ...hs],
                name: "ammonia",
                position: [...n.position],
            });
            ammonias.push(ammonia);
            this.molecules.push(ammonia);

            for (const h of hs) {
                n.bonds.push(h.atomId);
                h.bonds.push(n.atomId);
            }
        }

        return ammonias;
    }

    /**
     * Form an amino acid from available atoms.
     * Amino acids need: C, H, O, N (and sometimes S).
     * Simplified: requires 2C + 5H + 2O + 1N minimum (glycine).
     */
    formAminoAcid(aaType = "Gly") {
        const carbons = this.atomic.atoms.filter(
            a => a.atomicNumber === 6 && a.bonds.length === 0
        );
        const hydrogens = this.atomic.atoms.filter(
            a => a.atomicNumber === 1 && a.bonds.length === 0
        );
        const oxygens = this.atomic.atoms.filter(
            a => a.atomicNumber === 8 && a.bonds.length < 2
        );
        const nitrogens = this.atomic.atoms.filter(
            a => a.atomicNumber === 7 && a.bonds.length === 0
        );

        if (carbons.length < 2 || hydrogens.length < 5) return null;
        if (oxygens.length < 2 || nitrogens.length < 1) return null;

        const atoms = [];
        for (let i = 0; i < 2; i++) atoms.push(carbons.pop());
        for (let i = 0; i < 5; i++) atoms.push(hydrogens.pop());
        for (let i = 0; i < 2; i++) atoms.push(oxygens.pop());
        atoms.push(nitrogens.pop());

        const aa = new Molecule({
            atoms: atoms,
            name: aaType,
            position: [...atoms[0].position],
            isOrganic: true,
            functionalGroups: ["amino", "carboxyl"],
        });
        this.molecules.push(aa);
        this.aminoAcidCount++;

        // Form internal bonds
        for (let i = 1; i < atoms.length; i++) {
            atoms[0].bonds.push(atoms[i].atomId);
            atoms[i].bonds.push(atoms[0].atomId);
        }

        return aa;
    }

    /**
     * Form a nucleotide (sugar + phosphate + base).
     * Simplified: requires C5 + H10 + O7 + N2 + P minimum.
     * We use available atoms to approximate.
     */
    formNucleotide(base = "A") {
        const carbons = this.atomic.atoms.filter(
            a => a.atomicNumber === 6 && a.bonds.length === 0
        );
        const hydrogens = this.atomic.atoms.filter(
            a => a.atomicNumber === 1 && a.bonds.length === 0
        );
        const oxygens = this.atomic.atoms.filter(
            a => a.atomicNumber === 8 && a.bonds.length < 2
        );
        const nitrogens = this.atomic.atoms.filter(
            a => a.atomicNumber === 7 && a.bonds.length === 0
        );

        // Simplified requirements
        if (carbons.length < 5 || hydrogens.length < 8) return null;
        if (oxygens.length < 4 || nitrogens.length < 2) return null;

        const atoms = [];
        for (let i = 0; i < 5; i++) atoms.push(carbons.pop());
        for (let i = 0; i < 8; i++) atoms.push(hydrogens.pop());
        for (let i = 0; i < 4; i++) atoms.push(oxygens.pop());
        for (let i = 0; i < 2; i++) atoms.push(nitrogens.pop());

        const nuc = new Molecule({
            atoms: atoms,
            name: `nucleotide-${base}`,
            position: [...atoms[0].position],
            isOrganic: true,
            functionalGroups: ["sugar", "phosphate", "base"],
        });
        this.molecules.push(nuc);
        this.nucleotideCount++;

        for (let i = 1; i < atoms.length; i++) {
            atoms[0].bonds.push(atoms[i].atomId);
            atoms[i].bonds.push(atoms[0].atomId);
        }

        return nuc;
    }

    /** Run catalyzed reactions to form complex molecules. */
    catalyzedReaction(temperature, catalystPresent = false) {
        let formed = 0;
        const eaFactor = catalystPresent ? 0.3 : 1.0;
        const thermal = K_B * temperature;

        // Try to form amino acids
        if (thermal > 0 && this.atomic.atoms.length > 10) {
            const aaProb = Math.exp(-5.0 * eaFactor / (thermal + 1e-20));
            if (Math.random() < aaProb) {
                const aaType = AMINO_ACIDS[Math.floor(Math.random() * AMINO_ACIDS.length)];
                if (this.formAminoAcid(aaType)) {
                    formed++;
                    this.reactionsOccurred++;
                }
            }
        }

        // Try to form nucleotides
        if (thermal > 0 && this.atomic.atoms.length > 19) {
            const nucProb = Math.exp(-8.0 * eaFactor / (thermal + 1e-20));
            if (Math.random() < nucProb) {
                const bases = ["A", "T", "G", "C"];
                const base = bases[Math.floor(Math.random() * bases.length)];
                if (this.formNucleotide(base)) {
                    formed++;
                    this.reactionsOccurred++;
                }
            }
        }

        return formed;
    }

    /** Count molecules by type. */
    moleculeCensus() {
        const counts = {};
        for (const m of this.molecules) {
            const key = m.name || m.formula;
            counts[key] = (counts[key] || 0) + 1;
        }
        return counts;
    }

    toCompact() {
        const counts = this.moleculeCensus();
        const countStr = Object.entries(counts)
            .sort(([a], [b]) => a.localeCompare(b))
            .map(([k, v]) => `${k}:${v}`)
            .join(",");
        return (
            `CS[mol=${this.molecules.length} `
            + `H2O=${this.waterCount} `
            + `aa=${this.aminoAcidCount} `
            + `nuc=${this.nucleotideCount} `
            + `rxn=${this.reactionsOccurred} ${countStr}]`
        );
    }
}
